BCL bootstrap regressions introduced by CoreFX System import on AIX (regex, decimal)

## Steps to Reproduce

Build the runtime as normal.

### Current Behavior

```
MCS     [basic] mscorlib.dll
../../../external/corefx/src/Common/src/CoreLib/System/Collections/Generic/Dictionary.cs(231,22): warning CS0219: The variable `modified' is assigned but its value is never used
../../../external/corefx/src/Common/src/CoreLib/System/Collections/Generic/Dictionary.cs(240,18): warning CS0219: The variable `modified' is assigned but its value is never used
../../../external/corefx/src/Common/src/CoreLib/System/Collections/Generic/List.cs(950,21): warning CS0219: The variable `i' is assigned but its value is never used
../../../external/corefx/src/Common/src/CoreLib/System/Number.Formatting.cs(2125,17): warning CS0219: The variable `i' is assigned but its value is never used
../../../external/corefx/src/Common/src/CoreLib/System/Runtime/InteropServices/MemoryMarshal.cs(70,22): warning CS0219: The variable `localManager' is assigned but its value is never used
../../../external/corefx/src/Common/src/CoreLib/System/Runtime/InteropServices/MemoryMarshal.cs(89,22): warning CS0219: The variable `localManager' is assigned but its value is never used
../../../external/corefx/src/Common/src/CoreLib/System/Text/StringBuilder.cs(1760,17): warning CS0219: The variable `deltaLength' is assigned but its value is never used
../../../external/corefx/src/Common/src/CoreLib/System/Text/StringBuilder.cs(1798,25): warning CS0219: The variable `indexBeforeAdjustment' is assigned but its value is never used
../../../external/corefx/src/System.Memory/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Date.G.cs(42,19): warning CS0219: The variable `unused' is assigned but its value is never used
../../../external/corefx/src/System.Memory/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Date.O.cs(52,19): warning CS0219: The variable `unused' is assigned but its value is never used
../../../external/corefx/src/System.Memory/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Guid.cs(120,19): warning CS0219: The variable `unused' is assigned but its value is never used
../../../external/corefx/src/System.Memory/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Guid.cs(136,19): warning CS0219: The variable `unused' is assigned but its value is never used
../../../external/corefx/src/System.Memory/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Guid.cs(150,19): warning CS0219: The variable `unused' is assigned but its value is never used
../../../external/corefx/src/System.Memory/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Guid.cs(164,19): warning CS0219: The variable `unused' is assigned but its value is never used
../../../external/corefx/src/System.Memory/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Guid.cs(178,19): warning CS0219: The variable `unused' is assigned but its value is never used
../../../external/corert/src/System.Private.CoreLib/src/System/Threading/Tasks/FutureFactory.cs(551,26): warning CS0219: The variable `bWonSetException' is assigned but its value is never used
../../../external/corert/src/System.Private.CoreLib/src/System/Threading/Tasks/TaskScheduler.cs(384,53): warning CS0420: `System.Threading.Tasks.TaskScheduler.m_taskSchedulerId': A volatile field references will not be treated as volatile
System.Diagnostics.Tracing/EventSource.cs(161,15): warning CS3006: Overloaded method `System.Diagnostics.Tracing.EventSource.Write<T>(string, ref System.Diagnostics.Tracing.EventSourceOptions, ref T)' differing only in ref or out, or in array rank, is not CLS-compliant
System.Diagnostics.Tracing/EventSource.cs(156,15): (Location of the symbol related to previous warning)
corert/Decimal.cs(684,32): warning CS3006: Overloaded method `System.Decimal.Round(ref decimal, int, System.MidpointRounding)' differing only in ref or out, or in array rank, is not CLS-compliant
corert/Decimal.cs(682,31): (Location of the symbol related to previous warning)
corert/Decimal.cs(892,29): warning CS3006: Overloaded method `System.Decimal.Truncate(ref decimal)' differing only in ref or out, or in array rank, is not CLS-compliant
corert/Decimal.cs(885,31): (Location of the symbol related to previous warning)
../../../external/corefx/src/Common/src/CoreLib/System/Boolean.cs(29,22): warning CS0649: Field `System.Boolean.m_value' is never assigned to, and will always have its default value `false'
../../../external/corefx/src/System.Collections.NonGeneric/src/System/Collections/KeyValuePairs.cs(22,33): warning CS0414: The private field `System.Collections.KeyValuePairs._key' is assigned but its value is never used
../../../external/corefx/src/System.Collections.NonGeneric/src/System/Collections/KeyValuePairs.cs(25,33): warning CS0414: The private field `System.Collections.KeyValuePairs._value' is assigned but its value is never used
Compilation succeeded - 23 warning(s)

Unhandled Exception:
System.ArgumentException: parsing "^(?<flag>--|-|/)(?<name>[^:=]+)((?<sep>[:=])(?<value>.*))?$" - Invalid group name: Group names must begin with a word character.
  at System.Text.RegularExpressions.RegexParser.ScanGroupOpen () [0x00229] in <a906101b128e42548210c6ff25a3096c>:0
  at System.Text.RegularExpressions.RegexParser.ScanRegex () [0x001aa] in <a906101b128e42548210c6ff25a3096c>:0
  at System.Text.RegularExpressions.RegexParser.Parse (System.String re, System.Text.RegularExpressions.RegexOptions op) [0x00036] in <a906101b128e42548210c6ff25a3096c>:0
  at System.Text.RegularExpressions.Regex..ctor (System.String pattern, System.Text.RegularExpressions.RegexOptions options, System.TimeSpan matchTimeout, System.Boolean addToCache) [0x00097] in <a906101b128e42548210c6ff25a3096c>:0
  at System.Text.RegularExpressions.Regex..ctor (System.String pattern) [0x00000] in <a906101b128e42548210c6ff25a3096c>:0
  at Mono.Options.OptionSet..ctor (System.Converter`2[TInput,TOutput] localizer) [0x0000b] in <414bbd5b618b4b24a8858aaca22bc9da>:0
  at Mono.Options.OptionSet..ctor () [0x00000] in <414bbd5b618b4b24a8858aaca22bc9da>:0
  at Program.Main (System.String[] args) [0x00011] in <414bbd5b618b4b24a8858aaca22bc9da>:0
[ERROR] FATAL UNHANDLED EXCEPTION: System.ArgumentException: parsing "^(?<flag>--|-|/)(?<name>[^:=]+)((?<sep>[:=])(?<value>.*))?$" - Invalid group name: Group names must begin with a word character.
  at System.Text.RegularExpressions.RegexParser.ScanGroupOpen () [0x00229] in <a906101b128e42548210c6ff25a3096c>:0
  at System.Text.RegularExpressions.RegexParser.ScanRegex () [0x001aa] in <a906101b128e42548210c6ff25a3096c>:0
  at System.Text.RegularExpressions.RegexParser.Parse (System.String re, System.Text.RegularExpressions.RegexOptions op) [0x00036] in <a906101b128e42548210c6ff25a3096c>:0
  at System.Text.RegularExpressions.Regex..ctor (System.String pattern, System.Text.RegularExpressions.RegexOptions options, System.TimeSpan matchTimeout, System.Boolean addToCache) [0x00097] in <a906101b128e42548210c6ff25a3096c>:0
  at System.Text.RegularExpressions.Regex..ctor (System.String pattern) [0x00000] in <a906101b128e42548210c6ff25a3096c>:0
  at Mono.Options.OptionSet..ctor (System.Converter`2[TInput,TOutput] localizer) [0x0000b] in <414bbd5b618b4b24a8858aaca22bc9da>:0
  at Mono.Options.OptionSet..ctor () [0x00000] in <414bbd5b618b4b24a8858aaca22bc9da>:0
  at Program.Main (System.String[] args) [0x00011] in <414bbd5b618b4b24a8858aaca22bc9da>:0
../../build/library.make:352: recipe for target '../../class/lib/basic/mscorlib.dll' failed
gmake[8]: *** [../../class/lib/basic/mscorlib.dll] Error 1
```

I'm not sure why the regex parser is broken - I did some testing and it appears on a simple case of an IsMatch on `(?<name>.*)`, the regex on master throws up at RegexParser after a CaptureSlotFromName call, and the position seems to imply that its current position is at the n in name and is expecting a terminator, whereas the pre-breakage one has consumed the name and finds the terminator.

My suspicion is breakage somewhere in Char, maybe interaction with RegexCharClass? Unfortunately, since I can't bootstrap, I can only try to run binaries against monolite and view traces.

Since the file builds, unlike Jenkins, you can try to press on further until you hit [an issue on System.Security](https://gist.github.com/NattyNarwhal/8eb82ea33f0337d92e0f464afc42b83d).

### Expected Behavior

Compiles fine.

## On which platforms did you notice this

[ ] macOS
[ ] Linux
[ ] Windows
[X} AIX

My suspicion is that this is a general big endian issue, but I don't have access to another big endian platform to verify.

**Version Used**:

```
Mono JIT compiler version 5.17.0 (master/e09d690 Sat Jul 21 23:38:14  2018)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
        TLS:
        SIGSEGV:       normal
        Notification:  Thread + polling
        Architecture:  ppc
        Disabled:      none
        Misc:          softdebug
        Interpreter:   yes
        Suspend:       preemptive
        GC:            sgen
```

A diff against Mono CoreFX:

```diff
diff --git a/src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs b/src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs
index 971e2dd8f6..77ef4f1935 100644
--- a/src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs
+++ b/src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs
@@ -705,7 +705,9 @@ namespace System.Text.RegularExpressions
                 if (CharsRight() == 0)
                     break;

-                switch (ch = RightCharMoveRight())
+                ch = RightCharMoveRight();
+                System.Console.WriteLine(" ** RegexParser ** ch {0} len {1} pos {2}", ch, _pattern.Length, _currentPos);
+                switch (ch)
                 {
                     case ':':
                         // noncapturing group
@@ -774,8 +776,9 @@ namespace System.Text.RegularExpressions
                                         capnum = -1;

                                     // check if we have bogus characters after the number
-                                    if (CharsRight() > 0 && !(RightChar() == close || RightChar() == '-'))
-                                        throw MakeException(SR.InvalidGroupName);
+                                    if (CharsRight() > 0 && !(RightChar() == close || RightChar() == '-')) {
+                                        System.Console.WriteLine(" ** RegexParser ** Error #1: capnum {0}", capnum);
+                                        throw MakeException(SR.InvalidGroupName); }
                                     if (capnum == 0)
                                         throw MakeException(SR.CapnumNotZero);
                                 }
@@ -787,8 +790,9 @@ namespace System.Text.RegularExpressions
                                         capnum = CaptureSlotFromName(capname);

                                     // check if we have bogus character after the name
-                                    if (CharsRight() > 0 && !(RightChar() == close || RightChar() == '-'))
-                                        throw MakeException(SR.InvalidGroupName);
+                                    if (CharsRight() > 0 && !(RightChar() == close || RightChar() == '-')) {
+                                        System.Console.WriteLine(" ** RegexParser ** Error #2: capnum {0} capname {1} chars right {2} right char {3} close {4}", capnum, capname, CharsRight(), RightChar(), close);
+                                        throw MakeException(SR.InvalidGroupName); }
                                 }
                                 else if (ch == '-')
                                 {
@@ -797,6 +801,7 @@ namespace System.Text.RegularExpressions
                                 else
                                 {
                                     // bad group name - starts with something other than a word character and isn't a number
+                                    System.Console.WriteLine(" ** RegexParser ** Error #3: capnum {0}", capnum);
                                     throw MakeException(SR.InvalidGroupName);
                                 }

@@ -815,8 +820,9 @@ namespace System.Text.RegularExpressions
                                             throw MakeException(SR.Format(SR.UndefinedBackref, uncapnum));

                                         // check if we have bogus characters after the number
-                                        if (CharsRight() > 0 && RightChar() != close)
-                                            throw MakeException(SR.InvalidGroupName);
+                                        if (CharsRight() > 0 && RightChar() != close) {
+                                            System.Console.WriteLine(" ** RegexParser ** Error #4: capnum {0} uncapnum {1}", capnum, uncapnum);
+                                            throw MakeException(SR.InvalidGroupName); }
                                     }
                                     else if (RegexCharClass.IsWordChar(ch))
                                     {
@@ -828,12 +834,14 @@ namespace System.Text.RegularExpressions
                                             throw MakeException(SR.Format(SR.UndefinedNameRef, uncapname));

                                         // check if we have bogus character after the name
-                                        if (CharsRight() > 0 && RightChar() != close)
-                                            throw MakeException(SR.InvalidGroupName);
+                                        if (CharsRight() > 0 && RightChar() != close) {
+                                            System.Console.WriteLine(" ** RegexParser ** Error #5: capnum {0} uncapnum {1} uncapname {2}", capnum, uncapnum, uncapname);
+                                            throw MakeException(SR.InvalidGroupName); }
                                     }
                                     else
                                     {
                                         // bad group name - starts with something other than a word character and isn't a number
+                                        System.Console.WriteLine(" ** RegexParser ** Error #6");
                                         throw MakeException(SR.InvalidGroupName);
                                     }
                                 }
```

The [exception case #2](https://github.com/mono/corefx/blob/master/src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs#L791) is what's triggered by my test program, just something that runs IsMatch with `(?<name>.*)` on the first argument.

Output on aix/ppc64be, which has been broken by the PR:

```
aix$ MONO_PATH="mcs/class/lib/basic" runtime/mono-wrapper /home/calvin/regex-fucked.exe abcdef
 ** RegexParser ** ch < len 11 pos 3
 ** RegexParser ** Error #2: capnum 1 capname n chars right 7 right char a close >

Unhandled Exception:
System.ArgumentException: parsing "(?<name>.*)" - Invalid group name: Group names must begin with a word character.
  at System.Text.RegularExpressions.RegexParser.ScanGroupOpen () [0x002ad] in <9a5729ce8a4a49ff85197ddc732ac983>:0
  at System.Text.RegularExpressions.RegexParser.ScanRegex () [0x001aa] in <9a5729ce8a4a49ff85197ddc732ac983>:0
  at System.Text.RegularExpressions.RegexParser.Parse (System.String re, System.Text.RegularExpressions.RegexOptions op) [0x00036] in <9a5729ce8a4a49ff85197ddc732ac983>:0
  at System.Text.RegularExpressions.Regex..ctor (System.String pattern, System.Text.RegularExpressions.RegexOptions options, System.TimeSpan matchTimeout, System.Boolean addToCache) [0x00097] in <9a5729ce8a4a49ff85197ddc732ac983>:0
  at System.Text.RegularExpressions.Regex.IsMatch (System.String input, System.String pattern, System.Text.RegularExpressions.RegexOptions options, System.TimeSpan matchTimeout) [0x00000] in <9a5729ce8a4a49ff85197ddc732ac983>:0
  at System.Text.RegularExpressions.Regex.IsMatch (System.String input, System.String pattern) [0x00000] in <9a5729ce8a4a49ff85197ddc732ac983>:0
  at regexFucked.Main (System.String[] args) [0x00006] in <0ede8a6db4384b8598aacba63dcfb32d>:0
[ERROR] FATAL UNHANDLED EXCEPTION: System.ArgumentException: parsing "(?<name>.*)" - Invalid group name: Group names must begin with a word character.
  at System.Text.RegularExpressions.RegexParser.ScanGroupOpen () [0x002ad] in <9a5729ce8a4a49ff85197ddc732ac983>:0
  at System.Text.RegularExpressions.RegexParser.ScanRegex () [0x001aa] in <9a5729ce8a4a49ff85197ddc732ac983>:0
  at System.Text.RegularExpressions.RegexParser.Parse (System.String re, System.Text.RegularExpressions.RegexOptions op) [0x00036] in <9a5729ce8a4a49ff85197ddc732ac983>:0
  at System.Text.RegularExpressions.Regex..ctor (System.String pattern, System.Text.RegularExpressions.RegexOptions options, System.TimeSpan matchTimeout, System.Boolean addToCache) [0x00097] in <9a5729ce8a4a49ff85197ddc732ac983>:0
  at System.Text.RegularExpressions.Regex.IsMatch (System.String input, System.String pattern, System.Text.RegularExpressions.RegexOptions options, System.TimeSpan matchTimeout) [0x00000] in <9a5729ce8a4a49ff85197ddc732ac983>:0
  at System.Text.RegularExpressions.Regex.IsMatch (System.String input, System.String pattern) [0x00000] in <9a5729ce8a4a49ff85197ddc732ac983>:0
  at regexFucked.Main (System.String[] args) [0x00006] in <0ede8a6db4384b8598aacba63dcfb32d>:0
```

Output on FreeBSD/amd64, which works:

```
 ** RegexParser ** ch < len 11 pos 3
True
```

--- a/configure.ac
+++ b/configure.ac
@@ -40,7 +40,7 @@ MONO_VERSION_BUILD=`echo $VERSION | cut -d . -f 3`
 # This can be reset to 0 when Mono's version number is bumped
 # since it's part of the corlib version (the prefix '1' in the full
 # version number is to ensure the number isn't treated as octal in C)
-MONO_CORLIB_COUNTER=25
+MONO_CORLIB_COUNTER=26
 MONO_CORLIB_VERSION=`printf "1%02d%02d%02d%03d" $MONO_VERSION_MAJOR $MONO_VERSION_MINOR 0 $MONO_CORLIB_COUNTER`
 
 AC_DEFINE_UNQUOTED(MONO_CORLIB_VERSION,$MONO_CORLIB_VERSION,[Version of the corlib-runtime interface])

--- a/mcs/class/corlib/corlib.csproj
+++ b/mcs/class/corlib/corlib.csproj
@@ -294,7 +294,6 @@
     <Compile Include="..\..\..\external\corefx\src\Common\src\CoreLib\System\FormatException.cs" />
     <Compile Include="..\..\..\external\corefx\src\Common\src\CoreLib\System\FormattableString.cs" />
     <Compile Include="..\..\..\external\corefx\src\Common\src\CoreLib\System\Gen2GcCallback.cs" />
-    <Compile Include="..\..\..\external\corefx\src\Common\src\CoreLib\System\Globalization\CharUnicodeInfo.cs" />
     <Compile Include="..\..\..\external\corefx\src\Common\src\CoreLib\System\Globalization\CompareInfo.Invariant.cs" />
     <Compile Include="..\..\..\external\corefx\src\Common\src\CoreLib\System\Globalization\CompareInfo.cs" />
     <Compile Include="..\..\..\external\corefx\src\Common\src\CoreLib\System\Globalization\GlobalizationExtensions.cs" />
@@ -671,7 +670,6 @@
     <Compile Include="..\..\..\external\corert\src\System.Private.CoreLib\src\System\Array.cs" />
     <Compile Include="..\..\..\external\corert\src\System.Private.CoreLib\src\System\ByReference.cs" />
     <Compile Include="..\..\..\external\corert\src\System.Private.CoreLib\src\System\Collections\Generic\ArraySortHelper.cs" />
-    <Compile Include="..\..\..\external\corert\src\System.Private.CoreLib\src\System\Globalization\CharUnicodeInfoData.cs" />
     <Compile Include="..\..\..\external\corert\src\System.Private.CoreLib\src\System\Globalization\GlobalizationMode.cs" />
     <Compile Include="..\..\..\external\corert\src\System.Private.CoreLib\src\System\InsufficientMemoryException.cs" />
     <Compile Include="..\..\..\external\corert\src\System.Private.CoreLib\src\System\MissingFieldException.cs" />
@@ -791,6 +789,7 @@
     <Compile Include="..\referencesource\mscorlib\system\globalization\calendaralgorithmtype.cs" />
     <Compile Include="..\referencesource\mscorlib\system\globalization\calendardata.cs" />
     <Compile Include="..\referencesource\mscorlib\system\globalization\calendarweekrule.cs" />
+    <Compile Include="..\referencesource\mscorlib\system\globalization\charunicodeinfo.cs" />
     <Compile Include="..\referencesource\mscorlib\system\globalization\chineselunisolarcalendar.cs" />
     <Compile Include="..\referencesource\mscorlib\system\globalization\culturenotfoundexception.cs" />
     <Compile Include="..\referencesource\mscorlib\system\globalization\culturetypes.cs" />

--- a/mcs/class/corlib/corlib.dll.sources
+++ b/mcs/class/corlib/corlib.dll.sources
@@ -1076,8 +1076,10 @@ ReferenceSources/AppContextDefaultValues.cs
 ../referencesource/mscorlib/system/globalization/CalendricalCalculationsHelper.cs
 ../referencesource/mscorlib/system/globalization/calendardata.cs
 ../referencesource/mscorlib/system/globalization/calendarweekrule.cs
-../../../external/corefx/src/Common/src/CoreLib/System/Globalization/CharUnicodeInfo.cs
-../../../external/corert/src/System.Private.CoreLib/src/System/Globalization/CharUnicodeInfoData.cs
+../referencesource/mscorlib/system/globalization/charunicodeinfo.cs
+# TODO: make CoreFX's CharUnicodeInfo BE compatible
+#../../../external/corefx/src/Common/src/CoreLib/System/Globalization/CharUnicodeInfo.cs
+#../../../external/corert/src/System.Private.CoreLib/src/System/Globalization/CharUnicodeInfoData.cs
 ../referencesource/mscorlib/system/globalization/chineselunisolarcalendar.cs
 #../referencesource/mscorlib/system/globalization/compareinfo.cs
 ../referencesource/mscorlib/system/globalization/culturenotfoundexception.cs

--- a/mcs/class/referencesource/mscorlib/system/globalization/charunicodeinfo.cs
+++ b/mcs/class/referencesource/mscorlib/system/globalization/charunicodeinfo.cs
@@ -483,7 +483,9 @@ namespace System.Globalization {
             return (InternalGetUnicodeCategory(ch)) ;
         }
 
-        internal static UnicodeCategory GetUnicodeCategory(int ch) => GetUnicodeCategory((int)ch);
+#if MONO
+        public static UnicodeCategory GetUnicodeCategory(int ch) => GetUnicodeCategory((int)ch);
+#endif
 
         public static UnicodeCategory GetUnicodeCategory(String s, int index)
         {

--- a/mcs/class/referencesource/mscorlib/system/globalization/charunicodeinfo.cs
+++ b/mcs/class/referencesource/mscorlib/system/globalization/charunicodeinfo.cs
@@ -483,6 +483,8 @@ namespace System.Globalization {
             return (InternalGetUnicodeCategory(ch)) ;
         }
 
+        internal static UnicodeCategory GetUnicodeCategory(int ch) => GetUnicodeCategory((int)ch);
+
         public static UnicodeCategory GetUnicodeCategory(String s, int index)
         {
             if (s==null)

--- a/mcs/class/referencesource/mscorlib/system/globalization/charunicodeinfo.cs
+++ b/mcs/class/referencesource/mscorlib/system/globalization/charunicodeinfo.cs
@@ -484,7 +484,7 @@ namespace System.Globalization {
         }
 
 #if MONO
-        public static UnicodeCategory GetUnicodeCategory(int ch) => GetUnicodeCategory((char)ch);
+        public static UnicodeCategory GetUnicodeCategory(int codePoint) => GetUnicodeCategory((char)codePoint);
 #endif
 
         public static UnicodeCategory GetUnicodeCategory(String s, int index)

--- a/mcs/class/referencesource/mscorlib/system/globalization/charunicodeinfo.cs
+++ b/mcs/class/referencesource/mscorlib/system/globalization/charunicodeinfo.cs
@@ -484,7 +484,7 @@ namespace System.Globalization {
         }
 
 #if MONO
-        public static UnicodeCategory GetUnicodeCategory(int ch) => GetUnicodeCategory((int)ch);
+        public static UnicodeCategory GetUnicodeCategory(int ch) => GetUnicodeCategory((char)ch);
 #endif
 
         public static UnicodeCategory GetUnicodeCategory(String s, int index)

